src/storage-client/src/controller.rs

```diff
@@ -2714,14 +2736,25 @@ where
                         .expect("failed to connect")
                         .expect("failed to truncate write handle");
                 }
-                shard_id
+
+                // We only want to  clear shards from the finalization register
+                // if the since if actually empty.
+                if critical_since_handle.since().is_empty() {
+                    Some(shard_id)
+                } else {
+                    // TODO: periodically check shards that have not been successfully finalized.
```

Comment:
With the code as is, we already keep the shards that don't have `since == []`, and retry when this method is called again (when someone else tries to finalize), right?

Reply:
I'm not sure what "as is" means in this context. In the code submitted in the PR, we close the since, and remove it from the shard finalization WAL only if the since is closed.

Previously, we were removing shards from the WAL any time they were observed in this function, irrespective of the shard's observable since.

Reply:
Oh yes! I have all of this in another PR that's incoming––I was just trying to make this change as uncontroversial as possible. If you're OK with adding those extra bits in an additional PR, I can merge this and then get that up today/tomorrow.